### PR TITLE
Doesn't sign transaction #12

### PR DESCRIPTION
**Pull request by r8d8**

### CODE REVIEW

**Comment by sjmackenzie** [comment]
Rock on @r8d8 thanks!

**Comment by dulanov** [comment]
@r8d8 Every commit & PR should be presented as 'Problem > Solution'. It's very bad practice to make several commit with equals meaningless names!

**Comment by dulanov** [comment]
@r8d8 please fix build before

File: src/sign.rs [hunk 1]
```diff
@@ -0,0 +1,304 @@
+use tiny_keccak::Keccak;
+use secp256k1;
+use secp256k1::key::{SecretKey, PublicKey};
+use secp256k1::{Error, Message, RecoverableSignature, RecoveryId};
+use std::collections::HashMap;
```
Comment (dulanov) [inline src/sign.rs]: `use secp256k1::{self, ...}` instead of separate `use secp256k1;`

File: src/sign.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+use num_bigint::{BigUint};
+use rand::{Rng, thread_rng};
+
+lazy_static! {
+	pub static ref SECP256K1: secp256k1::Secp256k1 = secp256k1::Secp256k1::new();
+}
+
+pub trait Encodable {
+    fn encode(&self) -> Vec<u8>;
```
Comment (dulanov) [inline src/sign.rs]: `Into<Vec<u8>>`, why you need a new trait just to convert something?

File: src/sign.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+use secp256k1::{Error, Message, RecoverableSignature, RecoveryId};
+use std::collections::HashMap;
+use num_bigint::{BigUint};
+use rand::{Rng, thread_rng};
+
+lazy_static! {
+	pub static ref SECP256K1: secp256k1::Secp256k1 = secp256k1::Secp256k1::new();
+}
+
+pub trait Encodable {
```
Comment (dulanov) [inline src/sign.rs]: Every public elements (struct, trait, function) should have a documentation.

File: src/sign.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    fn encode(&self) -> Vec<u8>;
+}
+
+pub trait ToBytes {
```
Comment (dulanov) [inline src/sign.rs]: There is a `Into<T>` trait exactly to convert something to something, please don't invert new concepts.

File: src/sign.rs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    fn to_bytes(&self, out: &mut Vec<u8>);
+}
+
+impl ToBytes for usize {
```
Comment (dulanov) [inline src/sign.rs]: Here `usize`, next stay `as u64`, is it ok?

File: src/sign.rs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+
+        for i in 0..len {
+            let j = len - 1 - i;
+            out.push((*self >> (j * 8)) as u8);
+        }
+    }
+}
+
+///Signature for transaction.
+pub struct Signature {
```
Comment (dulanov) [inline src/sign.rs]: Please add a space after `///`

File: src/sign.rs [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    nonce: BigUint,
+    gas_price: BigUint,
+    gas_limit: BigUint,
+    to: BigUint,
+    value: BigUint,
+    pub data: Vec<u8>,
+}
+
+///Getter to extract first 32 bytes out of BigInt.
+macro_rules! impl_getter {
```
Comment (dulanov) [inline src/sign.rs]: Are you sure you need macros here instead of several simple methods?

File: src/sign.rs [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+macro_rules! extract_bigint {
+    ($args: expr, $val: expr) => {
+        BigUint::from_bytes_be($args.get($val).unwrap())
+    };
+}
+
+impl Transaction {
+
+    impl_getter!(nonce);
+    impl_getter!(gas_price);
```
Comment (dulanov) [inline src/sign.rs]: 'get' methods from old OOP school, i miss them too

File: src/sign.rs [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+    impl_getter!(value);
+
+    pub fn new(args: &HashMap<&str, Vec<u8>>) -> Self {
+        Transaction {
+            nonce: extract_bigint!(args, "nonce"),
+            gas_price: extract_bigint!(args, "gas_price"),
+            gas_limit: extract_bigint!(args, "gas_limit"),
+            to: extract_bigint!(args, "to"),
+            value: extract_bigint!(args, "value"),
+            data: args.get("data").unwrap().to_vec(),
```
Comment (dulanov) [inline src/sign.rs]: it's really crazy things, write a macro and use it from constructor!!!

File: src/sign.rs (hunk 4 above)
Comment (r8d8) [inline src/sign.rs]: `Into` can't be implemented for arbitrary types.
How one should handle this, without custom trait?

Comment (dulanov) [inline src/sign.rs]:
```
impl Into<Vec<u8>> for Transaction {
  ...
}
```

**Comment by splix** [comment]
Currently it doesn't support _EIP-155 Replay Protection_, but that fine actually, it's better to add it in a next iteration

**Comment by splix** [comment]
It will be really nice to have a real test for signature. I mean sign a transaction and validate actual binary data (bytes length, s/r/v bytes, etc). Right now I'm not sure if it signs at all

**Pull request closed without merging**